inst/dust/cows.cpp

```diff
@@ -205,15 +201,20 @@ class cows {
         const size_t n_herds_in_region = i_region_end - i_region_start;
         const size_t n_cows_in_region = internal.N[i_region_end - 1];
         const auto it_N = internal.N.begin() + i_region_start;
-        const size_t i_dst = std::distance(it_N, std::upper_bound(it_N, it_N + n_herds_in_region, u2 * n_cows_in_region));
+        const size_t i_dst = i_region_start + std::distance(it_N, std::upper_bound(it_N, it_N + n_herds_in_region, u2 * n_cows_in_region));
 
         const bool allow_movement = within_region || state_travel_allowed ||
-          mcstate::random::hypergeometric(rng_state, internal.export_I[i_src], export_N - internal.export_I[i_src], std::min(shared.n_test, internal.export_I[i_src])) == 0;
+          mcstate::random::hypergeometric(rng_state, internal.export_I[i_src], export_N - internal.export_I[i_src], std::min(shared.n_test, static_cast<real_type>(export_N))) == 0;
         if (allow_movement) {
           internal.import_S[i_dst] += internal.export_S[i_src];
           internal.import_E[i_dst] += internal.export_E[i_src];
           internal.import_I[i_dst] += internal.export_I[i_src];
           internal.import_R[i_dst] += internal.export_R[i_src];
+        } else {
+          internal.export_S[i_src] = 0;
+          internal.export_E[i_src] = 0;
+          internal.export_I[i_src] = 0;
+          internal.export_R[i_src] = 0;
```

Comment:
these lines are fairly harmless, but I think they aren't strictly needed as we do the `std::fill()` above?

Reply:
No I think they're actually crucial. It says that "if movement **isn't** allowed, then cancel the export". 
If these lines were removed, and movement was **not** allowed, then the import would not happen, but these exports would still get subtracted away.

Reply:
It's the addition of these lines that stops cows disappearing from the system at lines 223-227

Reply:
oh of course! Great work digging through this